verifyZeroInteractions guarantees zero interactions instead of just delegating to verifyNoMoreInteractions

ffissore opened the pull request:
Fixes #989

`verifyZeroInteractions` no longer delegates to `verifyNoMoreInteractions`
Instead, it fetches the number of invocations made to the mock and checks if
they are zero, failing otherwise

I'm not sure if the style of my code complies with mockito: please let me know what to change and where to look to make it more mockitoish

TimvdLippe commented:
You can always check if you comply with the code by running `./gradlew build`. That runs the test and checks the code style :)

ChristianSchwarz commented:
@ffissore 
Looks cool!  Can you add more tests?

szczepiq commented:
Thank you for submission! I have not looked into the implementation details, if nobody does picks up the review soon, I'll do it :)

Team, is the decision to change the behavior of "verifyZeroInteractions"? This is likely a disturbing change that will impact many projects migrating to Mockito v3. I'm +1 to this change but I have to acknowledge the impact ;)

TimvdLippe commented:
We would need to investigate the usage of this method. I think it is a logical change, but we have to prevent users from excessive code problems. They can of course always transfer everything to `verifyNoMoreInteractions` just before they are migrating, that shouldn't be a bit of a hassle.

ffissore commented:
@ChristianSchwarz I'll add more in the coming days
About the name: pls consider using a different name, `verifyNoInteractions` for example, and deprecating `verifyZeroInteractions` with an explanation

TimvdLippe commented on src/main/java/org/mockito/BDDMockito.java:
```diff
@@ -304,12 +304,22 @@ public T should(InOrder inOrder, VerificationMode mode) {
         /**
          * @see #verifyZeroInteractions(Object...)
          * @since 2.1.0
+         * @deprecated Please migrate your code to {@link #shouldHaveNoInteractions()}
```
Let's add a `@deprecated since 3.0.0` here, so we don't accidently remove it in the wrong version (which we did previously)

TimvdLippe commented on src/test/java/org/mockitousage/bugs/VerifyNoInteractionsIsLikeVerifyNeverTest.java:
```diff
@@ -0,0 +1,86 @@
+    verifyNoInteractions(collaborator);
+  }
+
+  @Test //(expected = NoInteractionsWanted.class)
```
Since we deprecate the method anyways, is this test really required?

ffissore commented:
As noted, in support to my proposal, I've changed my PR to include a new `verifyNoInteractions` and deprecated old `verifyZeroInteractions`. I've added some tests, and coverage will hopefully be happier.
@TimvdLippe I've added those "since 3.0.0" and removed my initial test class

TimvdLippe commented on src/main/java/org/mockito/Mockito.java:
```diff
@@ -2010,12 +2010,33 @@ public static void verifyNoMoreInteractions(Object... mocks) {
+     * <p>
+     * See examples in javadoc for {@link Mockito} class
+     *
+     * @param mocks to be verified
```
This still needs a `@since`

TimvdLippe commented on src/test/java/org/mockitousage/basicapi/ResetTest.java:
```diff
@@ -63,6 +63,13 @@ public void shouldRemoveAllInteractions() throws Exception {
     }
 
     @Test
+    public void shouldRemoveAllInteractions2() throws Exception {
```
I would rather not name it with `2`. Probably better as `should_remove_all_interactions_verify_no_interactions()`

TimvdLippe commented on src/test/java/org/mockitousage/basicapi/ResetTest.java:
```diff
@@ -79,6 +86,14 @@ public void shouldStubbingNotBeTreatedAsInteraction() {
     }
 
     @Test
+    public void shouldStubbingNotBeTreatedAsInteraction2() {
```
Same here

TimvdLippe commented on src/test/java/org/mockitousage/misuse/InvalidUsageTest.java:
```diff
@@ -104,6 +109,18 @@ public void shouldNotMockObjectMethodsOnInterface() throws Exception {
     }
 
+    @Test
+    public void shouldNotMockObjectMethodsOnInterface2() throws Exception {
```
Same here

TimvdLippe commented on src/test/java/org/mockitousage/misuse/InvalidUsageTest.java:
```diff
@@ -113,4 +130,15 @@ public void shouldNotMockObjectMethodsOnClass() throws Exception {
     }
+
+    @Test
+    public void shouldNotMockObjectMethodsOnClass2() throws Exception {
```
Same here

TimvdLippe commented on src/test/java/org/mockitousage/verification/NoMoreInteractionsVerificationTest.java:
```diff
@@ -126,6 +139,25 @@ public void shouldVerifyOneMockButFailOnOther() throws Exception {
     }
 
+    @Test
+    public void shouldVerifyOneMockButFailOnOther2() throws Exception {
```
Same here

TimvdLippe commented on src/test/java/org/mockitousage/stubbing/BasicStubbingTest.java:
```diff
@@ -67,6 +67,14 @@ public void should_stubbing_not_be_treated_as_interaction() {
     }
 
     @Test
+    public void should_stubbing_not_be_treated_as_interaction2() {
```
Same here

TimvdLippe commented on src/test/java/org/mockitousage/stubbing/BasicStubbingTest.java:
```diff
@@ -79,6 +87,18 @@ public void unfinished_stubbing_cleans_up_the_state() {
     }
 
     @Test
+    public void unfinished_stubbing_cleans_up_the_state2() {
```
Same here

ffissore commented:
Applied requested modifications and squashed commits

ffissore commented:
I don't know what went wrong on travis, tests pass on my machine using openjdk 8. Ping me if I can help

TimvdLippe commented:
These are flaky tests, so it seems fine :)

ChristianSchwarz commented on src/main/java/org/mockito/internal/MockitoCore.java:
```diff
@@ -137,6 +138,29 @@ public void verifyNoMoreInteractionsInOrder(List<Object> mocks, InOrderContext i
+                InvocationContainer invocations = getMockHandler(mock).getInvocationContainer();
+                VerificationDataImpl data = new VerificationDataImpl(invocations, null);
+                noInteractions().verify(data);
+            } catch (NotAMockException e) {
```
@ffissore We should avoid this catch block by calling `MockUtil.isMock(..)`

ffissore commented:
Remove try catch block
